packages/backend-app-api/src/services/implementations/discovery/MultipleBackendHostDiscovery.ts

```diff
@@ -0,0 +1,166 @@
+    });
+  }
+
+  constructor(options: {
+    instanceUrl: string;
+    gatewayUrl?: string;
+    rootFeatureRegistry: RootFeatureRegistryService;
+    discovery: DiscoveryService;
+  }) {
+    this.#gatewayUrl = options.gatewayUrl || options.instanceUrl;
+    this.#instanceUrl = options.instanceUrl;
+    this.#isGateway = !options.gatewayUrl;
+    this.#discovery = options.discovery;
+    this.#rootFeatureRegistry = options.rootFeatureRegistry;
+  }
+
+  async initialize() {
+    if (this.#isInitialized) {
+      throw new Error('Can not initialize twice.');
+    }
+    const features = await this.#rootFeatureRegistry.getFeatures();
+    const pluginIds = features
+      .filter(e => e.type === 'plugin')
+      .map(e => e.pluginId);
+    const plugins: PluginRegistrations = {};
+    for (const pluginId of pluginIds) {
+      plugins[pluginId] = {
+        internal: await this.#discovery.getBaseUrl(pluginId),
+        external: await this.#discovery.getExternalBaseUrl(pluginId),
+      };
+    }
+    this.addPlugins(this.#instanceUrl, plugins);
+  }
+
+  addPlugins(instanceUrl: string, plugins: PluginRegistrations) {
+    for (const [pluginId, urls] of Object.entries(plugins)) {
+      this.#plugins[pluginId] = urls;
+      if (!this.#instancePlugins[instanceUrl]) {
+        this.#instancePlugins[instanceUrl] = new Set();
+      }
+      this.#instancePlugins[instanceUrl].add(pluginId);
+    }
+  }
+
+  get plugins() {
+    return this.#plugins;
+  }
+
+  get instancePlugins() {
+    return this.#instancePlugins;
+  }
+
+  get isGateway() {
+    return this.#isGateway;
+  }
+
+  async getUrl(pluginId: string, key: 'external' | 'internal') {
+    // Because of how the services are initialized, lazy load the features when we first need them.
+    if (!this.#isInitialized) {
+      await this.initialize();
+      this.#isInitialized = true;
+    }
+    // If this instance knows about this plugin, return the value.
+    // The Gateway should have all plugins registered, individual instances will have just their
+    //  installed plugins.
+    if (this.#plugins[pluginId]) {
+      return this.#plugins[pluginId][key];
+    }
+    console.log(this.#isGateway, this.#plugins);
+    if (this.#isGateway) {
+      /**
+       * If we get here, either
+       *  a) we're requesting a route to a pluginId that hasn't been registered yet.
+       *  b) we're requesting a plugin that doesn't exist.
+       *
+       * Given the current design with a decentralized gateway definition, we don't know how many
+       *    nodes need to be registered. We could add a wait check here to allow for registration,
+       *    but this feels more like user error than anything, ie didn't configure backend A and
+       *    backend B correctly.
+       */
+    } else {
+      // As an instance plugin, fetch the installed plugins on the gateway URL.
+      const response = await fetch(
+        `${this.#gatewayUrl}/api/discovery/installed`,
```

Comment:
This should probably be a config defined value to support overridden prefixes on the router.